### DISCUSSION TITLE
Peer to Peer

### DISCUSSION THREAD

**Discussion by mcelrath** [braidpool/braidpool#65] [Ideas]
Since there's been several discussions about P2P technology choices including #57 #63 #64 I thought I'd try out this new GitHub Discussions page, since neither Issues nor Pull Requests are necessarily the right place to have such discussions, and it seems due to the lack of low-latency P2P libraries we will probably be discussing this for some time.

**Comment by ghost** [comment]
No need to wait for SV2 to stabilize, meanwhile all else can be hashed out, and different p2p impl. can be tried ...

As an example using a message.io (FramedTCP Transport `mio::net::{TcpListener, TcpStream};`) as [beacon] for discovery where [node]s p2p gossips after [peer-dictionary] is bootstrapped  --is a good start , then have a [coinbase-node] acting as a coinbase, accounting and measurement [node]. This would enable finding 'unknowns' in system fast. In practice would mean [node]s can hash with [mock-hasher]s to [node] that in turn can p2p to [DAG], here the gossip graph algorithm can be found / explored / settled.

The braidpool repo README.md says "For the moment, the braidpool node runs a simple p2p broadcast. " huh ...
Broadcast in what context ? not as tcp/ip broadcast, using tcp req-res is not broadcast.

 [braidcast] ing, should use streams ?.

The [invite] to join and get access to [beacon]s [peer-dictionary] can be a invite-code.onion => response would be a beacon that in turn  uses invite code as ID for new [node].

[invite] 
[beacon]
[peer-dictionary] 
[braidcast] [DAG]
[coinbase-node]
[node]
[mock-hasher]

**Reply by mcelrath** [reply]
Right now the code makes a TCP connection to itself and sends a "hello" message. There's no peer discovery yet. But obviously we need that.

**Comment by ghost** [comment]
1 GB same machine

Compiling message-io v0.18.0 (/opt/message-io)
    Finished release [optimized] target(s) in 2.81s
     Running `/opt/cargo-build/release/examples/throughput`
Sending 1073741824 bytes in chunks of 65507 bytes:

message-io Udp:         Throughput: 5.15 GB/s
message-io Tcp:         Throughput: 3.53 GB/s
message-io FramedTcp:   Throughput: 2.93 GB/s
message-io Ws:          Throughput: 1.94 GB/s

native Udp:             Throughput: 4.66 GB/s
native Tcp:             Throughput: 3.29 GB/s
native FramedTcp:       Throughput: 4.00 GB/s
native Ws:              Throughput: 1.92 GB/s

///!

1 MB same machine 

Sending 1048576 bytes in chunks of 65507 bytes:

message-io Udp:         Throughput: 2.81 GB/s
message-io Tcp:         Throughput: 2.77 GB/s
message-io FramedTcp:   Throughput: 1.70 GB/s
message-io Ws:          Throughput: 736.09 MB/s

native Udp:             Throughput: 3.83 GB/s
native Tcp:             Throughput: 3.02 GB/s
native FramedTcp:       Throughput: 2.26 GB/s
native Ws:              Throughput: 1.08 GB/s

**Reply by mcelrath** [reply]
We're looking for sending messages of a few kb, with as low latency at possible. We're more interested in latency than throughput.

Given that a FEC network library doesn't seem to exist, in the short term we should prioritize simplicity and upgradability.

**Comment by ghost** [comment]
![Braidcast](https://github.com/braidpool/braidpool/assets/127112857/9eee2b9c-f118-4856-8fe7-67cc7081d2f4)

**Reply by ghost** [reply]
[Braidcast] = System Message
[Beadcast] = Latency Priority Message

**Comment by ghost** [comment]
![Braidcast1](https://github.com/braidpool/braidpool/assets/127112857/1fa63fc8-9e05-4fcc-8f53-137ba1f1d9ad)

**Reply by ghost** [reply]
Better ? if [node] can receive aggregated Hashrate in [Bead] format directly ? ...

**Reply by mcelrath** [reply]
A bead is a single sha256d hash solution. The set of beads is the "aggregated" hashrate and all nodes receive all beads. They'll be a few kb in size.

It seems to me that we don't need two types of messages (system/latency priority). We're probably never going to care about throughput so should always use a latency optimized solution.

**Comment by ghost** [comment]
per [node][block][[num-shares-weight]][ts-start][ts-end] do You have gossip ... else fetch - verify ...

could be a control struct message after new tip to ensure each has correct DAG of last tip in epoch ...

**Comment by ghost** [comment]
"The Web Native File System (WNFS) is a distributed file system. It is versioned, logged, programmable, has strong-yet-flexible security, and is fully controlled by the end user. Service providers can validate writes without reading the contents of the file system, and minimal metadata is leaked."

"Nodes need to be encoded as dag-cbor."

https://github.com/wnfs-wg/

**Comment by ghost** [comment]
"Hash chains are a simple way to generate pseudorandom data, but are inefficient in situations that require long chains. This can cause unnecessary overhead for use cases including logical clocks, synchronizing the heads of a pseudorandom stream, or non-interactive key agreement. This paper presents the “skip ratchet”, a novel pseudorandom function that can be efficiently incremented by arbitrary intervals."

https://github.com/wnfs-wg/spec/blob/main/spec/skip-ratchet.md

https://eprint.iacr.org/2022/1078.pdf

https://github.com/wnfs-wg/rs-skip-ratchet

**Comment by ghost** [comment]
https://ipld.io/

**Comment by mcelrath** [comment]
@s2dd can you give some context as to how you think your linked ideas can be used in braidpool?

**Comment by ghost** [comment]
recall some snippet on matrix from @pool2win regarding distributed Filesystem and find it interesting

**Comment by ghost** [comment]
Range-Based Set Reconciliation

"Range-based set reconciliation is a simple approach to ef-
ficiently computing the union of two sets over a network,"

https://arxiv.org/pdf/2212.13567.pdf
